Add S3 support to sstables::test_env

xemul opened the pull request:
Currently there are only 2 tests for S3 -- the pure client test and compound object_store test that launches scylla, creates s3-backed table and CQL-queries it. At the same time there's a whole lot of small unit test for sstables functionality, part of it can run over S3 storage too.

This PR adds this support and patches several test cases to use it. More test cases are to come later on demand.

fixes: #13015

xemul commented:
upd:
- rebased
- made resharding test run over S3 too

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/799/

xemul commented:
Upd
- patched tests detected that s3_client sitting on sstables is not closed

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/858/

xemul commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/858/

#13211

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/862/

denesb commented on sstables/sstables.cc:
```diff
@@ -3471,6 +3475,7 @@ future<> sstable::destroy() {
     if (_cached_index_file) {
         co_await _cached_index_file->evict_gently();
     }
+    co_await _storage->destroy(*this);
```
An exception thrown from here can hide a previous error in `close_files()`. Speaking of which, that does sound like something specific to the local storage...

xemul replied:
Yes, that's true. However
1. filesystem_storage::destroy cannot return exceptional future
2. s3_storage::destroy calls s3_client.close() which in turn calls http::client::close() which in turn closes in and out for a socket and .close() for streams is :crossed_fingers: not supposed :four_leaf_clover: to return exceptions
3. clients live as "values" on sstables until #13458 , in which the plan is to move all clients somewhere else and stop making sstables close them upon destroy

xemul replied:
@denesb , do you agree with my justification of risky exception handling in this place? (not a merge ping, it has conflicts and the rebase is needed anyway)

denesb replied:
I'm fine with a "looks dangerous" but "works with the current code" as long as there is a plan to address this later, and as I understood this close will go away from sstable desctructors soon. So I'm fine with it.

xemul commented:
upd:
- rebased and fixed merge conflicts

One of the conflicts was the inability to use storage::get_stats() (removed by 2dbae856f84ebef8267e53ee542cc9aa1da27775) in the resharding test. Plug this case with a zero-size stub until #13713 settles, the test is useful not only due to its filter-size checks

avikivity commented on test/lib/test_services.cc:
```diff
@@ -164,13 +166,40 @@ test_env::impl::impl(test_env_config cfg)
+        auto wrap = std::make_shared<test_env_with_cql>(std::move(func), std::move(cfg));
+        auto db_cfg = make_shared<db::config>();
+        db_cfg->experimental_features({db::experimental_features_t::feature::KEYSPACE_STORAGE_OPTIONS});
+        return do_with_cql_env_thread([wrap = std::move(wrap)] (auto& cql_env) mutable {
```
I do think we need partial test_env for modularity. Each (well not each) module should provide a test_env that is suitable for testing its functionality and only starts its own dependencies.

denesb replied:
Good point.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1014/

xemul commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1014/

` topology_raft_disabled  debug  [ FAIL ] topology_raft_disabled.test_raft_upgrade.1 1231.44s`

Doesn't look related. I'll re-kick the CI and create an issue

nyh commented:
Seeing that this PR was already positively reviewed by two people, I merged it, but I want to raise one concern I have with one implication of this change.

If I understand correctly, after your patch some C++ test files, like test/boost/sstable_datafile_test.cc can no longer run independently (e.g., `build/dev/test/boost/sstable_datafile_test -- -c1 -m1G`) but must be run by `test.py` because it needs to run minio and set up all sorts of environment variables. What I would have preferred to see in this case (where minio is missing), that the s3 test be skipped rather than fail. It will allow developers to continue running individual tests without test.py (am I the only one still doing that?).

raphaelsc commented:
> Seeing that this PR was already positively reviewed by two people, I merged it, but I want to raise one concern I have with one implication of this change.
> 
> If I understand correctly, after your patch some C++ test files, like test/boost/sstable_datafile_test.cc can no longer run independently (e.g., `build/dev/test/boost/sstable_datafile_test -- -c1 -m1G`) but must be run by `test.py` because it needs to run minio and set up all sorts of environment variables. What I would have preferred to see in this case (where minio is missing), that the s3 test be skipped rather than fail. It will allow developers to continue running individual tests without test.py (am I the only one still doing that?).

I do that too. And it's important to keep this possibility.

@nyh @xemul I have an idea.

Not long time ago, I introduced test/lib/scylla_test_case, which essentially overrides the entry point for test cases, allowing for new command line options. See: scylla_tests_cmdline_options_processor::process_cmdline_options in test/lib/test_services.cc

What if we add a param for storage options and all the settings needed? So by default, sstable_datafile_test runs against local storage, and also s3 on demand.

Then test.py can easily run the same unmodified sstable_datafile_test, containing a single set of unit tests, with either local or s3 storage.

xemul commented:
>  It will allow developers to continue running individual tests without test.py (am I the only one still doing that?).

Yes, that'd be easy. I'll send a patch soon :)
